configuration/configurator/BuildSet.cpp

```diff
@@ -0,0 +1,246 @@
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "jptree.hpp"
+#include "XMLTags.h"
+#include "BuildSet.hpp"
+#include "SchemaCommon.hpp"
+
+using namespace CONFIGURATOR;
+
+#define LOOP_THRU_BUILD_SET for (int idx = 0; idx < m_buildSetArray.length(); idx++)
+
+static CBuildSetManager *s_pBuildSetManager = NULL;
+
+CBuildSetManager* CBuildSetManager::getInstance(const char* pBuildSetFile, const char* pBuildSetDirectory)
+{
+    if (s_pBuildSetManager == NULL)
+    {
+        s_pBuildSetManager = new CBuildSetManager();
+
+        if (pBuildSetFile != NULL)
+            s_pBuildSetManager->m_buildSetFile.set(pBuildSetFile);
+        else
+            s_pBuildSetManager->m_buildSetFile.set(DEFAULT_BUILD_SET_XML_FILE);
+
+        if (pBuildSetDirectory != NULL)
+            s_pBuildSetManager->m_buildSetDir.set(pBuildSetDirectory);
+        else
+            s_pBuildSetManager->m_buildSetDir.set(DEFAULT_BUILD_SET_DIRECTORY);
+
+        pBuildSetFile = s_pBuildSetManager->m_buildSetFile;
+        pBuildSetDirectory = s_pBuildSetManager->m_buildSetDir;
+
+        s_pBuildSetManager->m_buildSetPath.clear().appendf("%s%s%s", pBuildSetDirectory, pBuildSetDirectory[strlen(pBuildSetDirectory)-1] == '/' ? "" : "/", pBuildSetFile);
+        s_pBuildSetManager->populateBuildSet();
+    }
+    return s_pBuildSetManager;
+}
+
+CBuildSetManager::CBuildSetManager(const char* pBuildSetFile, const char* pBuildSetDir) : m_buildSetFile(pBuildSetFile), m_buildSetDir(pBuildSetDir)
+{
+    if (pBuildSetFile != NULL && pBuildSetDir != NULL)
+       m_buildSetPath.clear().appendf("%s%s%s", pBuildSetDir, pBuildSetDir[strlen(pBuildSetDir)-1] == '/' ? "" : "/", pBuildSetFile);
+}
+
+CBuildSetManager::CBuildSetManager()
+{
+}
+
+CBuildSetManager::~CBuildSetManager()
+{
+    m_buildSetArray.kill();
+}
+
+void CBuildSetManager::getBuildSetComponents(StringArray& buildSetArray) const
+{
+    int nLength = this->getBuildSetComponentCount();
+
+    for (int idx = 0; idx < nLength; idx++)
+    {
+        buildSetArray.append(this->getBuildSetComponentTypeName(idx));
+    }
+}
+
+void CBuildSetManager::getBuildSetServices(StringArray& buildSetArray) const
+{
+    int nLength = this->getBuildSetServiceCount();
+
+    for (int idx = 0; idx < nLength; idx++)
+    {
+        buildSetArray.append(this->getBuildSetServiceName(idx));
+    }
+}
+
+const char* CBuildSetManager::getBuildSetServiceName(int index) const
+{
+    return this->getBuildSetService(index)->getName();
+}
+
+const char* CBuildSetManager::getBuildSetServiceFileName(int index) const
+{
+    return this->getBuildSetService(index)->getSchema();
+}
+
+const char* CBuildSetManager::getBuildSetComponentTypeName(int index) const
+{
+    return this->getBuildSetComponent(index)->getName();
```

Comment:
do we need to ensure the index is in-bounds before attempting the getName() call?